File: vador/src/main/java/com/salesforce/vador/config/FilterDuplicatesConfig.java
/*******************************************************************************
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 ******************************************************************************/

/*******************************************************************************
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 ******************************************************************************/

package com.salesforce.vador.config;

import io.vavr.Function1;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Value;
import lombok.experimental.FieldDefaults;
import org.jetbrains.annotations.Nullable;

@Value
@FieldDefaults(level = AccessLevel.PACKAGE)
@Builder(buildMethodName = "prepare", builderMethodName = "toValidate", toBuilder = true)
public class FilterDuplicatesConfig<ValidatableT, FailureT> {
  // `andFailDuplicatesWith` is not mandatory for `findAndFilterDuplicatesWith`.
  // You may want to just filter without failing duplicates. So they are separated
  @Nullable Function1<ValidatableT, ?> findAndFilterDuplicatesWith;
  @Nullable FailureT andFailDuplicatesWith;
  @Nullable FailureT andFailNullKeysWith;
}

File: vador/src/main/kotlin/com/salesforce/vador/config/FilterDuplicatesConfig.kt
/**
 * ****************************************************************************
 * Copyright (c) 2022, salesforce.com, inc. All rights reserved. SPDX-License-Identifier:
 * BSD-3-Clause For full license text, see the LICENSE file in the repo root or
 * https://opensource.org/licenses/BSD-3-Clause
 * ****************************************************************************
 */
package com.salesforce.vador.config

import com.salesforce.vador.immutables.ConfigStyle
import io.vavr.Function1
import org.immutables.value.Value
import org.jetbrains.annotations.Nullable

@ConfigStyle
@Value.Immutable(copy = false)
internal abstract class AbstractFilterDuplicatesConfig<ValidatableT, FailureT> {
  // `andFailDuplicatesWith` is not mandatory for `findAndFilterDuplicatesWith`.
  // You may want to just filter without failing duplicates. So they are separated
  @get:Nullable abstract val findAndFilterDuplicatesWith: Function1<ValidatableT, *>?

  @get:Nullable abstract val andFailDuplicatesWith: FailureT?

  @get:Nullable abstract val andFailNullKeysWith: FailureT?
}
